### PR TITLE
Allow users to filter by tasks

### PR DESCRIPTION
**Pull request by daviddavis**
fixes #3144
https://pulp.plan.io/issues/3144

### CODE REVIEW

File: pulpcore/pulpcore/app/viewsets/custom_filters.py [hunk 1]
```diff
@@ -3,10 +3,37 @@
...
+        """
+
+        if not value:
+            return qs
+
+        model_class = self.model._meta.get_field(self.name).rel.to
+        instance = GenericNamedModelViewSet.get_resource(value, model_class)
+
+        key = "{}__pk".format(self.name)
+        return qs.filter(**{key: instance.pk})
```
Comment (dparalen) [inline pulpcore/pulpcore/app/viewsets/custom_filters.py]: I think getting the instance from the DB just to fetch its pk to construct a query is an overkill and might introduce performance issues; better might be to duplicate the URL resolving code from `GenericNamedModelViewSet.get_resource()`:
```python
match = resolve(urlparse(value).path)
key = "{}__pk".format(self.name)
return qs.filter(**{key: match.kwargs['pk']})
```

Even better might be to draw that code from `GenericNamedModelViewSet.get_resource()` into a new static method e.g `GenericNamedModelViewSet.get_view_by_url()` and use that instead

**Comment by daviddavis** [comment]
@dparalen thanks for the suggestion. I've updated the code.

I also had to catch `Exception` instead of an actual exception like ValidationError because it looks like if the UUID is malformed, the django exception handling code hits an exception:

https://gist.github.com/daviddavis/86ac89c3189839d0719f2f9865e258ab

Seems like a bug in django.

File: pulpcore/pulpcore/app/viewsets/custom_filters.py [hunk 2]
```diff
@@ -10,3 +13,31 @@ class CharInFilter(filters.BaseInFilter, filters.CharFilter):
...
+            Queryset filtered by the foreign key instance
+        """
+
+        if not value:
+            return qs
+
+        try:
+            match = resolve(urlparse(value).path)
+        except Exception:
+            return qs
```
Comment (dparalen) [inline pulpcore/pulpcore/app/viewsets/custom_filters.py]: I'm a bit worried this might, in case one (by mistake) specifies a malformed URL, not limit the returned list of objects (tasks): `GET@/v3/tasks/?worker=this_aint_url&state__in=waiting` might return more tasks than expected. Shouldn't in such a case a 400 response be generated instead? Or a 404?

File: pulpcore/pulpcore/app/viewsets/custom_filters.py [hunk 3]
```diff
@@ -10,3 +13,31 @@ class CharInFilter(filters.BaseInFilter, filters.CharFilter):
...
+        Args:
+            qs (django.db.models.query.QuerySet): The Queryset to filter
+            value (string): href representing an instance of foreign key
+
+        Returns:
+            Queryset filtered by the foreign key instance
+        """
+
+        if not value:
+            return qs
```
Comment (dparalen) [inline pulpcore/pulpcore/app/viewsets/custom_filters.py]: this seems now obsoleted by line 39 but I'm wondering the same: shouldn't a 400/404 be raised in case empty value in here?

File: pulpcore/pulpcore/app/viewsets/custom_filters.py [hunk 4]
```diff
@@ -1,12 +1,53 @@
...
+                                              format(name=self.name))
+
+        try:
+            match = resolve(urlparse(value).path)
+        except Resolver404:
+            raise serializers.ValidationError(detail=_('URI not valid: {u}').format(u=value))
+
+        pk = match.kwargs['pk']
+        try:
+            UUID(pk, version=4)
```
Comment (daviddavis) [inline pulpcore/pulpcore/app/viewsets/custom_filters.py]: I am manually validating UUID here otherwise it gets validated much later on when the query gets executed.

Not sure if this is the best way to validate a UUID though.

**Comment by daviddavis** [comment]
@dparalen updated.

File: pulpcore/pulpcore/app/viewsets/custom_filters.py (hunk 4 above)
Comment (dparalen) [reply pulpcore/pulpcore/app/viewsets/custom_filters.py]: I think this is the best we can do; another PR might follow up by making this value-url-pk workflow some sort of an utility function (to reuse e.g in the `GenericNamedModelViewSet.get_resource()`) but this is cool too.

Btw I'd say this may be worth unit testing for edge conditions.

**Comment by dparalen** [comment]
Just by looking this seems LGTM; I'd like to play with it a bit in the console; will do a bit later.
If anyone else can do it faster, feel free to approve!

**Comment by dralley** [comment]
I played around with this locally and everything seems functional.  Could you file an issue for pulp-smash to test this?

File: pulpcore/pulpcore/app/viewsets/custom_filters.py [hunk 5]
```diff
@@ -1,12 +1,53 @@
...
+    """
+
+    def filter(self, qs, value):
+        """
+        Args:
+            qs (django.db.models.query.QuerySet): The Queryset to filter
+            value (string): href representing an instance of foreign key
+
+        Returns:
+            Queryset filtered by the foreign key instance
```
Comment (dralley) [inline pulpcore/pulpcore/app/viewsets/custom_filters.py]: Paste the type down here for the return as well.

Comment (dralley) [inline pulpcore/pulpcore/app/viewsets/custom_filters.py]: Could you paste the queryset type down here as well?

**Comment by daviddavis** [comment]
@dralley updated the pulp-smash issue and docstring 👍

**Comment by dralley** [comment]
@daviddavis , approved :+1: 

I like @dparalen 's suggestion of having a reusable utility function, but instead of doing that now let's file a refactor issue.  Adding unit testing can be done then.

**Comment by daviddavis** [comment]
So I opened this issue to refactor: https://pulp.plan.io/issues/3471

And I outlined some edge cases in the pulp-smash test.

File: pulpcore/pulpcore/app/viewsets/task.py [hunk 6]
```diff
@@ -12,10 +13,19 @@
 
 
 class TaskFilter(filterset.FilterSet):
+    state_in_list = CharInFilter(name='state', lookup_expr='in')
+    worker = HyperlinkRelatedFilter(name='worker')
```
Comment (dparalen) [inline pulpcore/pulpcore/app/viewsets/task.py]: I checked docs generated for `GET@/api/v3/tasks/` but I can't see neither type nor description of that field. I'm not able to figure out how to get that though (`label='text'` doesn't work) so I'm cool ;)

Comment (daviddavis) [reply pulpcore/pulpcore/app/viewsets/task.py]: I looked into this but the solutions seem painful. It would be nice to have an easy way to add descriptions for filters though and maybe something we could look into in the future.
